add `extern "custom"` functions

folkertdev opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/140770)*
<!-- homu-ignore:end -->

tracking issue: rust-lang/rust#140829
previous discussion: https://github.com/rust-lang/rust/issues/140566

In short, an `extern "custom"` function is a function with a custom ABI, that rust does not know about. Therefore, such functions can only be defined with `#[unsafe(naked)]` and `naked_asm!`, or via an `extern "C" { /* ... */ }` block. These functions cannot be called using normal rust syntax: calling them can only be done from inline assembly.

The motivation is low-level scenarios where a custom calling convention is used. Currently, we often pick `extern "C"`, but that is a lie because the function does not actually respect the C calling convention.

At the moment `"custom"` seems to be the name with the most support. That name is not final, but we need to pick something to actually implement this.

r? @traviscross 
cc @tgross35 

try-job: x86_64-apple-2

folkertdev committed:
add `extern "custom"` functions

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14892610806/job/41828192489) [(plain)](https://github.com/rust-lang/rust/commit/f61cadaccf6493357a79a7881e191531910f081f/checks/41828192489/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.0s done
#18 DONE 32.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0004]: non-exhaustive patterns: `Conv::Custom` not covered
   --> compiler/rustc_codegen_cranelift/src/abi/mod.rs:45:11
    |
45  |     match c {
    |           ^ pattern `Conv::Custom` not covered
    |
note: `Conv` defined here
   --> /checkout/compiler/rustc_target/src/callconv/mod.rs:532:1
    |
532 | pub enum Conv {
```

</details>

bjorn3 commented on compiler/rustc_smir/src/stable_mir/abi.rs:
```diff
@@ -430,6 +430,8 @@ pub enum CallConvention {
     PreserveMost,
     PreserveAll,
 
+    Custom,
```
Not directly related to this PR, but why is this enum not marked as `#[non_exhaustive]`?

folkertdev replied:
Yeah I was thinking the same. How can this be "stable mir" when we have to make breaking changes like this.

tgross35 commented on compiler/rustc_passes/src/custom_abi.rs:
```diff
@@ -0,0 +1,153 @@
+            _ => None,
+        };
+
+        if let Some((span, ExternAbi::Custom)) = opt_span_and_abi {
+            self.tcx.dcx().emit_err(AbiCustomCall { span });
+        }
```
From https://github.com/rust-lang/rust/issues/140566#issuecomment-2846205457 and the few messages following, it sounds like there are a couple of other ABIs that should be uncallable (specifically interrupt ABIs). Maybe this should go via an `is_abi_callable` function next to `is_abi_supported` in `rustc_target`?

mejrs commented on compiler/rustc_passes/src/custom_abi.rs:
```diff
@@ -0,0 +1,153 @@
+                    && let hir::ForeignItemKind::Fn(sig, _, _) = foreign_item.kind
+                    && sig.header.is_safe()
+                {
+                    let len = "safe ".len() as u32;
+                    let safe_span = sig.span.shrink_to_lo().with_hi(sig.span.lo() + BytePos(len));
```
This feels somewhat brittle. `rustc_ast::ast::Safety` has the span for this, if you can get that here. Or perhaps this entire pass should be in `rustc_ast_passes` instead, similar to the `MissingAbi` lint/errors?

tgross35 replied:
If the decision is to reject function params and return types (TBD at https://github.com/rust-lang/rust/issues/140566#issuecomment-2860606615, not sure what your preference is), that could probably also be part of the AST pass.

folkertdev replied:
it is fragile, but kind of needed for good suggestions. The `rustc_ast::ast::Safety` span is just the keyword, it's missing the space that also must be deleted, and there is no good way of getting the span of what follows as far as I can see.

Moving this logic over to `rustc_ast_passes` does work nicely.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14894306554/job/41833638071) [(plain)](https://github.com/rust-lang/rust/commit/93971b9a70e5fe7b5ac572496435c915b5d1aff6/checks/41833638071/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.9s done
#18 DONE 32.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
To only update this specific test, also pass `--test-args print-calling-conventions.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/print-calling-conventions.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/print-calling-conventions" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--print" "calling-conventions"
--- stdout -------------------------------
C
C-cmse-nonsecure-call
C-cmse-nonsecure-entry
C-unwind
```

</details>

tgross35 commented on tests/ui/abi/bad-custom.rs:
```diff
@@ -0,0 +1,73 @@
+//@ check-fail
+//@ only-x86_64
```
This probably doesn't need to be restricted since the asm is empty (and the assembler shouldn't see this anyway)

asquared31415 replied:
I believe we have a `needs-asm` test command or something like it to ensure that the test is only run when `asm!` can be used.

folkertdev replied:
yeah I'm using `//@ needs-asm-support`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14910498314/job/41883443145) [(plain)](https://github.com/rust-lang/rust/commit/a72606d15716d332b90c9e56792e031127a09459/checks/41883443145/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 19.9s done
#18 DONE 26.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 4.879
[RUSTC-TIMING] test test:false 8.135
    Finished `release` profile [optimized] target(s) in 1m 10s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 70.112
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.701
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.655
    Finished `release` profile [optimized] target(s) in 3.83s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.176
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.162
---
---- extern_abi::tests::guarantee_lexicographic_ordering stdout ----

thread 'extern_abi::tests::guarantee_lexicographic_ordering' panicked at compiler/rustc_abi/src/extern_abi/tests.rs:30:5:
assertion `left == right` failed
  left: [C { unwind: false }, CCmseNonSecureCall, CCmseNonSecureEntry, C { unwind: true }, Rust, Aapcs { unwind: false }, Aapcs { unwind: true }, AvrInterrupt, AvrNonBlockingInterrupt, Cdecl { unwind: false }, Cdecl { unwind: true }, EfiApi, Fastcall { unwind: false }, Fastcall { unwind: true }, GpuKernel, Msp430Interrupt, PtxKernel, RiscvInterruptM, RiscvInterruptS, RustCall, RustCold, Stdcall { unwind: false }, Stdcall { unwind: true }, System { unwind: false }, System { unwind: true }, SysV64 { unwind: false }, SysV64 { unwind: true }, Thiscall { unwind: false }, Thiscall { unwind: true }, Unadjusted, Vectorcall { unwind: false }, Vectorcall { unwind: true }, Win64 { unwind: false }, Win64 { unwind: true }, X86Interrupt, Custom]
 right: [C { unwind: false }, CCmseNonSecureCall, CCmseNonSecureEntry, C { unwind: true }, Rust, Aapcs { unwind: false }, Aapcs { unwind: true }, AvrInterrupt, AvrNonBlockingInterrupt, Cdecl { unwind: false }, Cdecl { unwind: true }, Custom, EfiApi, Fastcall { unwind: false }, Fastcall { unwind: true }, GpuKernel, Msp430Interrupt, PtxKernel, RiscvInterruptM, RiscvInterruptS, RustCall, RustCold, Stdcall { unwind: false }, Stdcall { unwind: true }, System { unwind: false }, System { unwind: true }, SysV64 { unwind: false }, SysV64 { unwind: true }, Thiscall { unwind: false }, Thiscall { unwind: true }, Unadjusted, Vectorcall { unwind: false }, Vectorcall { unwind: true }, Win64 { unwind: false }, Win64 { unwind: true }, X86Interrupt]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    extern_abi::tests::guarantee_lexicographic_ordering

test result: FAILED. 4 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 318.07µs

error: test failed, to rerun pass `-p rustc_abi --lib`
Build completed unsuccessfully in 0:29:26
  local time: Thu May  8 16:18:00 UTC 2025
  network time: Thu, 08 May 2025 16:18:01 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14911955127/job/41888348253) [(plain)](https://github.com/rust-lang/rust/commit/64868d3d1d2180deabbd369002e32400535a0e0d/checks/41888348253/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 19.8s done
#19 DONE 25.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
[RUSTC-TIMING] proc_macro test:false 4.967
[RUSTC-TIMING] test test:false 8.269
    Finished `release` profile [optimized] target(s) in 1m 10s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 70.586
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.710
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.670
    Finished `release` profile [optimized] target(s) in 3.79s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/ci-gcc/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.178
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.162
---
---- extern_abi::tests::guarantee_lexicographic_ordering stdout ----

thread 'extern_abi::tests::guarantee_lexicographic_ordering' panicked at compiler/rustc_abi/src/extern_abi/tests.rs:30:5:
assertion `left == right` failed
  left: [C { unwind: false }, CCmseNonSecureCall, CCmseNonSecureEntry, C { unwind: true }, Rust, Aapcs { unwind: false }, Aapcs { unwind: true }, AvrInterrupt, AvrNonBlockingInterrupt, Cdecl { unwind: false }, Cdecl { unwind: true }, EfiApi, Fastcall { unwind: false }, Fastcall { unwind: true }, GpuKernel, Msp430Interrupt, PtxKernel, RiscvInterruptM, RiscvInterruptS, RustCall, RustCold, Stdcall { unwind: false }, Stdcall { unwind: true }, System { unwind: false }, System { unwind: true }, SysV64 { unwind: false }, SysV64 { unwind: true }, Thiscall { unwind: false }, Thiscall { unwind: true }, Unadjusted, Vectorcall { unwind: false }, Vectorcall { unwind: true }, Win64 { unwind: false }, Win64 { unwind: true }, X86Interrupt, Custom]
 right: [C { unwind: false }, CCmseNonSecureCall, CCmseNonSecureEntry, C { unwind: true }, Rust, Aapcs { unwind: false }, Aapcs { unwind: true }, AvrInterrupt, AvrNonBlockingInterrupt, Cdecl { unwind: false }, Cdecl { unwind: true }, Custom, EfiApi, Fastcall { unwind: false }, Fastcall { unwind: true }, GpuKernel, Msp430Interrupt, PtxKernel, RiscvInterruptM, RiscvInterruptS, RustCall, RustCold, Stdcall { unwind: false }, Stdcall { unwind: true }, System { unwind: false }, System { unwind: true }, SysV64 { unwind: false }, SysV64 { unwind: true }, Thiscall { unwind: false }, Thiscall { unwind: true }, Unadjusted, Vectorcall { unwind: false }, Vectorcall { unwind: true }, Win64 { unwind: false }, Win64 { unwind: true }, X86Interrupt]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    extern_abi::tests::guarantee_lexicographic_ordering

test result: FAILED. 4 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 352.74µs

error: test failed, to rerun pass `-p rustc_abi --lib`
Build completed unsuccessfully in 0:29:06
  local time: Thu May  8 17:37:37 UTC 2025
  network time: Thu, 08 May 2025 17:37:37 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bjorn3 commented:
> Emit some extern "custom" errors in ast_validation:wq

Did you forget to press escape before quitting vim?

folkertdev commented:
ah, yeah. Well it'll probably get squashed before this is ready

folkertdev commented:
I've made a proper tracking issue at https://github.com/rust-lang/rust/issues/140829

bors commented:
:umbrella: The latest upstream changes (presumably #141002) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #141984) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

folkertdev commented:
I've rebased and changed the implementation to match https://github.com/rust-lang/rust/pull/141774, which is both shorter and apparently faster too.

bors commented:
:umbrella: The latest upstream changes (presumably #142099) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #141774) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

workingjubilee commented:
I believe T-lang has already basically okayed this experiment per [TC's comment here](https://github.com/rust-lang/rust/issues/140566#issuecomment-2861617481) and I think any concerns that TC winds up having can be addressed in followup. It's not like lang's gonna FCP this tomorrow, is it? Rolling a compiler review.

r? compiler

workingjubilee commented on compiler/rustc_abi/src/canon_abi.rs:
```diff
@@ -79,6 +82,7 @@ impl fmt::Display for CanonAbi {
                 X86Call::Vectorcall => ExternAbi::Vectorcall { unwind: false },
                 X86Call::Win64 => ExternAbi::Win64 { unwind: false },
             },
+            CanonAbi::Custom => ExternAbi::Custom,
```
Up?

ABIs valid on all targets should always precede all other ABIs. Same place in the ordering you put the variant in originally is fine. I realize that the ExternAbi listing isn't well-ordered but I'd like to preserve this much.

workingjubilee replied:
Thanks!

workingjubilee commented on compiler/rustc_ast_passes/src/errors.rs:
```diff
@@ -824,3 +824,51 @@ pub(crate) struct MissingAbi {
+
+    #[suggestion(
+        ast_passes_suggestion,
+        applicability = "maybe-incorrect",
```
I think this should be machine-applicable? The only reason it might not be, I guess, is if we think that `extern "custom"` was incorrectly applied, but they had to write it first in order for any sort of mechanical tooling to apply the suggestion.

workingjubilee commented on compiler/rustc_ast_passes/src/errors.rs:
```diff
@@ -824,3 +824,51 @@ pub(crate) struct MissingAbi {
+
+    #[suggestion(
+        ast_passes_suggestion,
+        applicability = "maybe-incorrect",
```
This should be machine-applicable for similar reasons?

workingjubilee commented on compiler/rustc_codegen_cranelift/src/abi/mod.rs:
```diff
@@ -64,6 +64,11 @@ pub(crate) fn conv_to_call_conv(
             unreachable!("tried to use {c:?} call conv which only exists on an unsupported target")
         }
+
+        // Functions with this calling convention can only be called from assembly, but it is
+        // possible to declare an `extern "custom"` block, so the backend still needs a calling
+        // convention for declaring foreign functions.
+        CanonAbi::Custom => default_call_conv,
```
Up, please?

workingjubilee commented on compiler/rustc_codegen_gcc/src/abi.rs:
```diff
@@ -273,6 +273,10 @@ pub fn conv_to_fn_attribute<'gcc>(conv: CanonAbi, arch: &str) -> Option<FnAttrib
             X86Call::SysV64 => FnAttribute::X86SysvAbi,
             X86Call::Win64 => FnAttribute::X86MsAbi,
         },
+        // Functions with this calling convention can only be called from assembly, but it is
+        // possible to declare an `extern "custom"` block, so the backend still needs a calling
+        // convention for declaring foreign functions.
+        CanonAbi::Custom => return None,
```
Please.

workingjubilee commented on compiler/rustc_codegen_llvm/src/abi.rs:
```diff
@@ -677,6 +677,10 @@ impl llvm::CallConv {
                 X86Call::Vectorcall => llvm::X86_VectorCall,
                 X86Call::Win64 => llvm::X86_64_Win64,
             },
+            // Functions with this calling convention can only be called from assembly, but it is
+            // possible to declare an `extern "custom"` block, so the backend still needs a calling
+            // convention for declaring foreign functions.
+            CanonAbi::Custom => llvm::CCallConv,
```
Por favor?

workingjubilee commented on compiler/rustc_target/src/spec/abi_map.rs:
```diff
@@ -153,6 +153,7 @@ impl AbiMap {
                 | ExternAbi::X86Interrupt,
                 _,
             ) => return AbiMapping::Invalid,
+            (ExternAbi::Custom, _) => CanonAbi::Custom,
```
Yep.

workingjubilee reviewed (changes requested):
I'm sorry but I have a huge and tedious nit that should be a tidy rule but isn't because Reasons.

bors commented:
:umbrella: The latest upstream changes (presumably #142181) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #141435) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented on compiler/rustc_codegen_cranelift/src/abi/mod.rs:
```diff
@@ -51,6 +51,11 @@ pub(crate) fn conv_to_call_conv(
         CanonAbi::Rust | CanonAbi::C => default_call_conv,
         CanonAbi::RustCold => CallConv::Cold,
 
+        // Functions with this calling convention can only be called from assembly, but it is
+        // possible to declare an `extern "custom"` block, so the backend still needs a calling
+        // convention for declaring foreign functions.
+        CanonAbi::Custom => default_call_conv,
```
This seems reasonable but I wonder if down the line clif (or the other backends) might want to have some kind of calling convention representation that can be `bug!`ed on if it gets directly called somewhere. (cc @bjorn3)

tgross35 commented on tests/ui/abi/custom.rs:
```diff
@@ -0,0 +1,91 @@
+    unsafe { asm!("call {}", sym increment, inout("rax") x) };
+    assert_eq!(x, 42);
+
+    let mut x: u32;
+    unsafe { asm!("call {}", sym Thing::is_even, inout("rax") 42 => x) };
+    assert!(x != 0);
```
It should be zero-extended anyway but would specifying an `al` clobber with `x: u8` be more accurate here?

tgross35 commented on compiler/rustc_hir_typeck/src/callee.rs:
```diff
@@ -135,6 +138,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         output
     }
 
+    /// Functions of type `extern "custom" fn(/* ... */)` cannot be called using `ExprKind::Call`.
+    /// These functions have a calling convention that is unknown to rust, hence it cannot generate
+    /// code for the call. The only way to execute such a function is via inline assembly.
```
Minor nit, newline after the first sentence to split the summary

tgross35 commented on compiler/rustc_hir_typeck/src/callee.rs:
```diff
@@ -135,6 +138,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+        let abi = match callee_ty.kind() {
+            ty::FnDef(def_id, _) => self.tcx.fn_sig(def_id).skip_binder().skip_binder().abi,
+            ty::FnPtr(_, header) => header.abi,
+            _ => return,
```
Are there other valid callees? If so, it may be good to `return` from them specifically then `bug!` on the fallback `_`.

folkertdev replied:
Hmm, that raises the question of what `async extern "custom"  fn foo() {}` should do. It always returns `impl Future`, so I think we should just disallow this.

tgross35 replied:
Agreed to disallowing async 👍 that's likely pretty tied to the Rust ABI too.

folkertdev replied:
I believe that what can happen here is that the scrutinee is some `impl Fn()` (or similar) value: it is callable, but not a function definition or function pointer. 

@oli-obk you've been working in this area a bunch: do you have advice on getting the ABI here (anything that implicitly uses the `extern "Rust"` (closures, `Fn()`) is irrelevant, we're just looking for callable items with the `"custom"` ABI. I suspect `&unsafe extern "custom" fn()` in particular is going to be tricky and is missed by the current logic. Is there some standard way to get at the "real" type?

oli-obk replied:
autoderef has already happened at this point. You can invoke `autoderef.final_ty()` to get the actual type that we'll be invoking things on. That should resolve any concerns about `&T`. Since Nothing but function pointers and function definitions is callable *and* can have a fn abi other than `"Rust"`, the current code otherwise seems fine. Some tests are good regardless. So basically just ensuring that you can't use the "custom" abi in impl methods or trait methods and calling references to function pointers or even function definitions.

folkertdev replied:
Ah, that's helpful! I'm still hitting `ty::Alias` and `ty::Adt` in this match though (maybe these become errors later). Can we safely ignore those cases with the `_ => return`?

oli-obk replied:
Yea there is no way to make them have any abi but Rust

tgross35 commented on compiler/rustc_hir_analysis/messages.ftl:
```diff
@@ -1,3 +1,7 @@
+hir_analysis_abi_custom_clothed_function =
+    functions with the `"custom"` ABI must be naked
+    .suggestion = add the `#[unsafe(naked)]` attribute to this function
```
I think it's worth hinting at the other way to have a `custom` function. Not the best wording, but maybe something like:

```text
    items with the `"custom"` ABI can only be declared externally or defined via naked functions
    .suggestion = convert this to an `#[unsafe(naked)]` function
```

tgross35 reviewed (commented):
This looks pretty clean to me, and well tested. Left a handful of small comments but nothing major.

folkertdev commented on tests/ui/abi/bad-custom.rs:
```diff
@@ -0,0 +1,92 @@
+//@ check-fail
```
It's tricky, because some of the fixes introduce issues elsewhere (e.g. we remove all arguments from the signature, but now call sites run into a type error. I've demoted that lint back to `maybe-incorrect`.


Also some lints fight, e.g. 

```rust 
impl Negate for Thing {
    extern "custom" fn negate(a: i64) -> i64 {
        -a
    }
}
``` 

becomes

```rust
impl Negate for Thing {
    unsafe #[unsafe(naked)]
extern "custom" fn negate() {
        -a
    }
}
```

so yeah, I'm not really sure what to do about that.

tgross35 replied:
Huh, interesting. Maybe I'd propose that:

* Changing safety should be machine-applicable
* Removing args/ret could be machine-applicable if the function is already naked, meaning the author can't already be making use of the arguments. Not sure if we have an easy way to classify that "if"
* Adding `#[unsafe(naked)]` isn't machine-applicable since that won't work unless they intended it to be naked. In fact, maybe we should just suggest rewriting as a naked function but don't suggest adding `#[unsafe(naked)]`? Since only adding the attribute probably isn't enough.

    If the body only contains `naked_asm!(...)` then I think we could give a machine-applicable `#[unsafe(naked)]` suggestion, but that would be a more general change rather than something related to the ABI here.

Cc @compiler-errors in case you have any thoughts about these compiler errors 🙂

folkertdev replied:
> Changing safety should be machine-applicable

I don't think that works with `pub`, so I've demoted it. 

> Removing args/ret could be machine-applicable if the function is already naked, meaning the author can't already be making use of the arguments. Not sure if we have an easy way to classify that "if"

We don't have a great way of figuring that out. That lint runs in `rustc_ast_passes` which is neat but doesn't give convenient access to attributes.

> Adding `#[unsafe(naked)]` isn't machine-applicable

Agreed, I've changed that. I think suggesting `#[unsafe(naked)]` is still nice though, the errors do sort of guide you through the process. 

---

So, in summary, only removing the `safe` keyword is machine-applicable right now. I think that's OK for this feature, at least for now. Given the target audience for this ABI, I don't think fixing the errors we report is really going to cause confusion.

tgross35 replied:
Sounds reasonable to me, +1 for getting this in without worrying too much about autofix.

tgross35 commented on tests/ui/abi/bad-custom.rs:
Based on https://github.com/rust-lang/rust/pull/140770#discussion_r2137433750 could you add a test for an `async` function and a `const` function? Const is rejected for calling asm (will probably always be that way) and `async` seems to be rejected by checks on the form it desugars to, but it would be good to make sure these don't regress.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=2823536b9788410319e9f67b66a480a1

folkertdev replied:
We now reject `async` explicitly, and `const` is rejected because of the `naked_asm!`

folkertdev replied:
I also added some additonal tests for references and `impl Fn`.

bors commented:
:umbrella: The latest upstream changes (presumably #141883) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
Looks great with green CI and a squash. Thanks for picking this up!

@bors delegate+

bors commented:
:v: @folkertdev, you can now approve this pull request!

If @tgross35 told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@tgross35`
<!-- homu: {"type":"Delegated","delegator":"tgross35","delegate":"folkertdev"} -->

bors commented:
:pushpin: Commit 22de67f9e152c768cceda092b2a689c994e86908 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 22de67f9e152c768cceda092b2a689c994e86908 -->
<!-- homu: {"type":"Approved","sha":"22de67f9e152c768cceda092b2a689c994e86908","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/142409#issuecomment-2967425586

bors commented:
:hourglass: Trying commit 635195230f15e9b13b1fc37fbdb7bb8ef86de96e with merge 83eaabcd225185dd1c7c0373ea8c209967ea06dd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"635195230f15e9b13b1fc37fbdb7bb8ef86de96e","merge_sha":"83eaabcd225185dd1c7c0373ea8c209967ea06dd"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15618369484/job/44001687976)
Build commit: 33aabc703ff3ae8fd55fe25c96b5461fb8601fca (`33aabc703ff3ae8fd55fe25c96b5461fb8601fca`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15618369484/job/44001687976"},"merge_sha":"33aabc703ff3ae8fd55fe25c96b5461fb8601fca"} -->

folkertdev commented:
Looks like the job that failed with the npm stuff got re-run? In any case, it's green now, the try build worked, so

@bors r=@tgross35

bors commented:
:pushpin: Commit 5f73ce2b7ebfc0234a4bc736217110a2af1e2380 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 5f73ce2b7ebfc0234a4bc736217110a2af1e2380 -->
<!-- homu: {"type":"Approved","sha":"5f73ce2b7ebfc0234a4bc736217110a2af1e2380","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
@rust-timer build 2bf564a237a7f8c4f7586a356ef57958d84e7c1a

(Testing https://github.com/rust-lang/rust/pull/142443)

rust-timer commented:
Queued 2bf564a237a7f8c4f7586a356ef57958d84e7c1a with parent 0d6ab209c525e276cbe7544cbd39a3c3619b6b18, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0d6ab209c525e276cbe7544cbd39a3c3619b6b18&end=2bf564a237a7f8c4f7586a356ef57958d84e7c1a).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->
